### PR TITLE
Add Stella model support

### PR DESCRIPTION
**Pull request by farshidz**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Add support for Stella 400m text embedding models (https://huggingface.co/dunzhang/stella_en_400M_v5)

* **What is the current behavior?** (You can also link to an open issue here)
Not supported

* **What is the new behavior (if this is a feature change)?**
New model type supported. Type is `hf_stella` as these need to be loaded slightly differently compared to other HF models.

Note we assume the linear layer has been absorbed into the model, which is the case the Marqo version of the base model, and is expected from fine-tuned models as well.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
Yes

* **Related Python client changes** (link commit/PR here)
No

* **Related documentation changes** (link commit/PR here)
Pending

* **Other information**:

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/core/inference/embedding_models/hugging_face_model.py [hunk 1]
```diff
@@ -35,11 +46,21 @@ def __init__(self, model_properties: dict, device: str, model_auth: Optional[Mod
...
-            return HuggingFaceModelProperties(**model_properties)
+            parsed_properties = HuggingFaceModelProperties(**model_properties)
         except ValidationError as e:
             raise InvalidModelPropertiesError(f"Invalid model properties: {model_properties}. Original error {e}") \
                 from e
 
+        if self._model_flags.trust_remote_code or self._tokenizer_flags.trust_remote_code:
+            if not parsed_properties.trust_remote_code:
+                raise InvalidModelPropertiesError(
+                    f"The specified model requires the 'trust_remote_code' attribute to be set to True. "
```
Comment (wanliAlex) [inline src/marqo/core/inference/embedding_models/hugging_face_model.py]: nit: trustRemoteCode

File: src/marqo/core/inference/embedding_models/hugging_face_model.py [hunk 2]
```diff
@@ -119,8 +146,12 @@ def _load_from_zip_file(self) -> Tuple:
 
         model_dir = self.extract_huggingface_archive(zip_file_path)
         try:
-            model = AutoModel.from_pretrained(model_dir).to(self.device)
-            tokenizer = AutoTokenizer.from_pretrained(model_dir)
+            model = AutoModel.from_pretrained(model_dir).to(
```
Comment (wanliAlex) [inline src/marqo/core/inference/embedding_models/hugging_face_model.py]: typo here? **kwargs should be in side from_pretrained?

**Review by wanliAlex — CHANGES_REQUESTED** [review]

File: src/marqo/core/inference/embedding_models/hugging_face_model.py (hunk 2 above)
Comment (farshidz) [reply src/marqo/core/inference/embedding_models/hugging_face_model.py]: Addressed

File: src/marqo/core/inference/embedding_models/hugging_face_model.py (hunk 1 above)
Comment (farshidz) [reply src/marqo/core/inference/embedding_models/hugging_face_model.py]: Addressed

**Review by wanliAlex — DISMISSED** [review]

**Review by wanliAlex — APPROVED** [review]

**Comment by tomaarsen** [comment]
Hello!

Out of curiosity, why do you rely on Transformers and your own/your users' configuration regarding e.g. pooling methods, dimensions, etc. when this information is stored in the model itself in the Sentence Transformers configuration.

I have a feeling like you can fully replace https://github.com/marqo-ai/marqo/blob/mainline/src/marqo/core/inference/embedding_models/hugging_face_model.py with a class like 
```python
from sentence_transformers import SentenceTransformer

class HuggingFaceModel(AbstractEmbeddingModel):
    def __init__(
            self,
            model_properties: dict,
            device: str,
            model_auth: Optional[ModelAuth] = None,
            model_flags: Optional[HuggingFaceModelFlags] = None,
            tokenizer_flags: Optional[HuggingFaceTokenizerFlags] = None
    ):
        super().__init__(model_properties, device, model_auth)
        self.model_properties = self._build_model_properties(model_properties)
        self._model = None

    def _load_necessary_components(self):
        if self.model_properties.name:
            model_name_or_path = self.model_properties.name
        if self.model_properties.url or (self.model_properties.model_location and self.model_properties.model_location.s3):
            zip_file_path = download_model(
                repo_location=self.model_properties.model_location,
                url=self.model_properties.url,
                auth=self.model_auth,
                download_dir=ModelCache.hf_cache_path
            )
            model_name_or_path = self.extract_huggingface_archive(zip_file_path)

        self._model = SentenceTransformer(
            model_name_or_path,
            token=self.model_auth.hf.token if self.model_auth and self.model_auth.hf else None,
            trust_remote_code=self.model_properties.trust_remote_code,
        )
        # Bonus: You can also add `truncate_dim`, i.e. https://huggingface.co/blog/matryoshka support
        # Bonus 2: You can also add `backend`, i.e. ONNX and OpenVINO support: https://sbert.net/docs/sentence_transformer/usage/efficiency.html

    def encode(self, sentence: Union[str, List[str]], normalize=True, **kwargs)
        return self._model.encode(sentence, normalize_embeddings=normalize, **kwargs)
```

This will give you immediate support for all of these models: https://huggingface.co/models?library=sentence-transformers&sort=trending
Even if they have odd requirements, `trust_remote_code`, custom modules, unusual pooling approaches, prompt prefixes, etc., or even non-`AutoModel` based embedding solutions like static embeddings: https://huggingface.co/tomaarsen/static-bert-uncased-gooaq.

Disclaimer: I maintain Sentence Transformers

cc @wanliAlex @farshidz

- Tom Aarsen

**Comment by wanliAlex** [comment]
> Hello!
> 
> Out of curiosity, why do you rely on Transformers and your own/your users' configuration regarding e.g. pooling methods, dimensions, etc. when this information is stored in the model itself in the Sentence Transformers configuration.
> 
> I have a feeling like you can fully replace https://github.com/marqo-ai/marqo/blob/mainline/src/marqo/core/inference/embedding_models/hugging_face_model.py with a class like
> 
> ```python
> from sentence_transformers import SentenceTransformer
> 
> class HuggingFaceModel(AbstractEmbeddingModel):
>     def __init__(
>             self,
>             model_properties: dict,
>             device: str,
>             model_auth: Optional[ModelAuth] = None,
>             model_flags: Optional[HuggingFaceModelFlags] = None,
>             tokenizer_flags: Optional[HuggingFaceTokenizerFlags] = None
>     ):
>         super().__init__(model_properties, device, model_auth)
>         self.model_properties = self._build_model_properties(model_properties)
>         self._model = None
> 
>     def _load_necessary_components(self):
>         if self.model_properties.name:
>             model_name_or_path = self.model_properties.name
>         if self.model_properties.url or (self.model_properties.model_location and self.model_properties.model_location.s3):
>             zip_file_path = download_model(
>                 repo_location=self.model_properties.model_location,
>                 url=self.model_properties.url,
>                 auth=self.model_auth,
>                 download_dir=ModelCache.hf_cache_path
>             )
>             model_name_or_path = self.extract_huggingface_archive(zip_file_path)
> 
>         self._model = SentenceTransformer(
>             model_name_or_path,
>             token=self.model_auth.hf.token if self.model_auth and self.model_auth.hf else None,
>             trust_remote_code=self.model_properties.trust_remote_code,
>         )
>         # Bonus: You can also add `truncate_dim`, i.e. https://huggingface.co/blog/matryoshka support
>         # Bonus 2: You can also add `backend`, i.e. ONNX and OpenVINO support: https://sbert.net/docs/sentence_transformer/usage/efficiency.html
> 
>     def encode(self, sentence: Union[str, List[str]], normalize=True, **kwargs)
>         return self._model.encode(sentence, normalize_embeddings=normalize, **kwargs)
> ```
> 
> This will give you immediate support for all of these models: https://huggingface.co/models?library=sentence-transformers&sort=trending Even if they have odd requirements, `trust_remote_code`, custom modules, unusual pooling approaches, prompt prefixes, etc., or even non-`AutoModel` based embedding solutions like static embeddings: https://huggingface.co/tomaarsen/static-bert-uncased-gooaq.
> 
> Disclaimer: I maintain Sentence Transformers
> 
> cc @wanliAlex @farshidz
> 
> * Tom Aarsen

Hey Tom, 

Thanks a lot for your suggestions and the code snippet. Yes, we also realized that `transformers` is harder to use and we do have a plan to wrap those models into a `sentence-transformer` based class. 

We really appreciate the insight – it’s super helpful to get input directly from the maintainer!

Best regards,
Li

**Pull request merged**